cli/lsp/config.rs

```diff
@@ -139,25 +145,178 @@ impl WorkspaceSettings {
   }
 }
 
+#[derive(Debug, Clone, Default)]
+pub struct ConfigSnapshot {
+  pub client_capabilities: ClientCapabilities,
+  pub root_uri: Option<Url>,
+  pub settings: Settings,
+}
+
+impl ConfigSnapshot {
+  pub fn specifier_enabled(&self, specifier: &ModuleSpecifier) -> bool {
+    if let Some(settings) = self.settings.specifiers.get(specifier) {
+      settings.1.enable
+    } else {
+      self.settings.workspace.enable
+    }
+  }
+}
+
+enum ConfigRequest {
+  Specifier(ModuleSpecifier, ModuleSpecifier),
+  Workspace,
+}
+
 #[derive(Debug, Default, Clone)]
+pub struct Settings {
+  pub specifiers:
+    HashMap<ModuleSpecifier, (ModuleSpecifier, SpecifierSettings)>,
+  pub workspace: WorkspaceSettings,
+}
+
+#[derive(Debug)]
 pub struct Config {
   pub client_capabilities: ClientCapabilities,
   pub root_uri: Option<Url>,
-  pub specifier_settings: HashMap<ModuleSpecifier, SpecifierSettings>,
-  pub workspace_settings: WorkspaceSettings,
+  settings: Arc<Mutex<Settings>>,
+  tx: mpsc::UnboundedSender<ConfigRequest>,
```

Comment:
You're probably using an unbounded queue so you don't have to .await sends? I don't see anything really wrong with that except that I don't really have a feeling for how many requests can get queued up.

Reply:
I am doing it because that is what we did in the DiagnosticServer, versus an informed decision. The most likely case of a queue of requests would be the situation where you are doing a multi-file file and replace (which is where the deadlock was occurring before) where multiple `didOpen` notifications are sent from the client, which now requires a request back from the server to the client to get the configuration of a resource it hasn't seen before.

Reply:
Ok, I moved over to a `channel` versus an `unbound_channel` and put a good guess in there of what the max queue should look like.